### PR TITLE
add sunburst to pie chart

### LINKED ISSUE #8563: Multiple breakout pie chart
metabase is awesome.

i love this project. i consider transfer my dashboard chart to metabase. and i got some trouble.

is metabase have pie chart like this?

wx20180920-163838 2x

here is my "works".

wx20180920-164619 2x

the english is not my first language.  maybe have some grammar error.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### LINKED ISSUE #47209: [Epic] Multi-dimensional / Sunburst Pie chart
**Links**
- product doc: [Offer new viz type: Multi-dimensional / Sunburst Pie chart](https://www.notion.so/metabase/Offer-new-viz-type-Multi-dimensional-Sunburst-Pie-chart-1bf26a2d04064f45ba8918c8f87d15ac)
- feature branch:  add-sunburst-to-pie-chart
- pr: https://github.com/metabase/metabase/pull/47208

Total estimate: 20 days / 4 weeks

### ECharts Data Layer (`chartModel`) [1-2 days]
- [x] add middle ring slices to `chartModel.slices`
- [x] render basic middle ring in option
- [x] add outer ring slices to `chartModel.slices`
- [x] render basic outer ring in option
- [x] ~update `isNonPositive` to account for middle and outer ring slices~
- [x] only show unaggregated warning if last ring has unaggregated values
- [x] aggregate slices into `other` in middle and outer rings

### ECharts Viz Layer (option object, rendering viz settings, styles)

**Labels** [2-3 days]
- [x] conditionally combine slice name with percent in label
    - [x] add viz setting for this, compute a sensible default
- [x] update label rotation (might depend on ring level)
- [x] re-introduce dimension formatting to echarts code for middle & outer slices
- [x] dynamic font sizing (depends on canvas size, rings)
- [x] update `getIsLabelVisible` to account for label rotation, number of rings

**Colors** [1 day]
- [x] get light/dark variants from brand color to use in middle & outer slices
- [x] ~alternative logic for white-labeled / embedded colors?~

**Styles** [1-2 days]
- [x] update inner/outer radius logic (depends on rings)
- [x] apply minimum slice size adjust to middle & outer rings
- [x] update slice border width logic

**Total** [1-2 days]
- [x] only show sub-total for sub-slice on hover
- [x] ~dynamic font size depending on rings + canvas size~

**Tooltip/Hover/Click** [2-3 days]
- [x] update hover chart events so tooltip knows which slices to use
- [x] update tooltip UI
- [x] update legend hover events
- [x] update click event to return correct value
- [x] update tooltip UI for "other" slices

### Viz Settings Sidebar UI [3 days]
- [x] move measure to top
- [x] create parent component to handle all of `pie.dimension`, `pie.middle_dimension`, `pie.outer_dimension`, `pie.rows`, & `pie.sort_rows` settings
    - [x] render dropdown menus for selecting `pie.dimension`, `pie.middle_dimension`, `pie.outer_dimension`
        - [x] render `Add ring` button and modify settings/UI appropriately
    - [x] rerender/re-wire `ChartSettingOrderedSimple` and `SliceNameWidget` components for `pie.rows` setting 
    - [x] use dnd-kit to support re-ordering rings, update settings appropriately
    
### Viz Settings Computations  [1 day]
- [x] compute suitable defaults based on query results, re-use in static viz
- [x] update definition of when pie/sunburst is sensible (collab with Conor on this)

### Testing [2-3 days]
- [x] loki static viz tests
- [x] e2e tests

### Bugs
- [x] #47869
- [x] #47776
- [x] #47867
- [x] #47775
- [x] #47931
- [x] field name remapping from admin settings doesn’t work for non-inner slice
- [x] metric value should be aggregated for click actions
- [x] minimum slice percentage setting should be based on parent slice
- [x] total value text should be truncated or hidden
- [x] ~latitude/longitude don't work as metric values~

### PR DESCRIPTION
**Pull request by EmmadUsmani**
Closes https://github.com/metabase/metabase/issues/47209
Closes https://github.com/metabase/metabase/issues/8563

### Description

Adds a sunburst feature to the pie chart, using up to three dimension columns.

### How to verify

1. Create a question with three dimensions, one metric, see example below

Screenshot 2024-09-26 at 2 48 39 PM

2. Visualize as pie chart, should have automatically select dimension and metric columns

### Demo

Screenshot 2024-09-26 at 2 49 14 PM

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

### CODE REVIEW

**Comment by EmmadUsmani** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/metabase/metabase/47208?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#47208** <a href="https://app.graphite.dev/github/pr/metabase/metabase/47208?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#46607** <a href="https://app.graphite.dev/github/pr/metabase/metabase/46607?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46505** <a href="https://app.graphite.dev/github/pr/metabase/metabase/46505?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#46466** <a href="https://app.graphite.dev/github/pr/metabase/metabase/46466?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EmmadUsmani and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by cdeweyx — APPROVED** [review]
Approved! Just a few things to follow-up on:
- Drillthrough on "Other" segment on inner ring not working
- Getting the same color assigned next to eachother most of the time ([this discussion](https://metaboat.slack.com/archives/C064QMXEV9N/p1727288761644389?thread_ts=1727282563.519899&cid=C064QMXEV9N)) which makes the chart harder to parse. Nice to have depending on effort, not required to merge, since my understanding is that the color assignment logic isn't the easiest to work with (happy to be wrong about that though).

**Comment by EmmadUsmani** [comment]
@cdeweyx Thanks for the quick review and feedback!

- Drill through not being available on the "other" slice is intentional, since we haven't had drills available on that slice before on the pie chart. We could add it in the future, although the functionality would likely be limited to just `Filter by this value`, since we can't easily write an MBQL query that returns the rows for the other slice.

- Yeah improving the color logic would require some investigation. I think I'll focus on merging the sunburst as is first, and if I have time next week I can look into this.

**Comment by cdeweyx** [comment]
> @cdeweyx Thanks for the quick review and feedback!
> 
> * Drill through not being available on the "other" slice is intentional, since we haven't had drills available on that slice before on the pie chart. We could add it in the future, although the functionality would likely be limited to just `Filter by this value`, since we can't easily write an MBQL query that returns the rows for the other slice.
> * Yeah improving the color logic would require some investigation. I think I'll focus on merging the sunburst as is first, and if I have time next week I can look into this.

Gotcha, both sound reasonable to me. Let's merge this 👍

**Review by kdoh — COMMENTED** [review]
One tiny thing - if I have a sunburst with only two rings, it feels strange to call the outer one - "middle"
Screenshot 2024-10-01 at 11 12 06 AM

File: frontend/src/metabase/visualizations/echarts/pie/model/index.ts [hunk 1]
```diff
@@ -123,108 +258,286 @@ export function getPieChartModel(
...
+
+      // Create or update node for middle dimension
+      const middleDimensionKey = getKeyFromDimensionValue(
+        row[colDescs.middleDimensionDesc.index],
+      );
+      let middleDimensionNode = dimensionNode.children.get(
+        String(middleDimensionKey),
+      );
+
+      if (middleDimensionNode == null) {
```
Comment (alxnddr) [inline frontend/src/metabase/visualizations/echarts/pie/model/index.ts:363]: This middle/outer dimension node logic looks quite similar, would it be possible to extract a function for node creation with shared logic?

File: .loki/reference/chrome_laptop_static_viz_PieChart_All_Negative.png
[the remark this answers is not in the document]
Comment (EmmadUsmani) [reply .loki/reference/chrome_laptop_static_viz_PieChart_All_Negative.png:1]: It should be empty, we decided to change this behavior https://metaboat.slack.com/archives/C064QMXEV9N/p1726783722560789

File: frontend/src/metabase/visualizations/components/ChartTooltip/EChartsTooltip/EChartsTooltip.tsx [hunk 2]
```diff
@@ -68,9 +69,9 @@ export const EChartsTooltip = ({
         <tbody>
           {paddedRows.map((row, index) => {
             return !row.isSecondary ? (
-              <TooltipRow key={index} {...row} />
+              <TooltipRow key={String(index)} {...row} />
```
[the remark this answers is not in the document]
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/components/ChartTooltip/EChartsTooltip/EChartsTooltip.tsx]: I removed the key

File: frontend/src/metabase/visualizations/echarts/pie/model/index.ts (hunk 1 above)
Comment (EmmadUsmani) [reply frontend/src/metabase/visualizations/echarts/pie/model/index.ts]: Yup, just did that 8dc70e4e8bbf81f4aee936ee6d18d236b3d63a31

**Review by alxnddr — APPROVED** [review]
Went through it another time, works great.
Can you please add a bit more specs around the click behavior? Can be done separately

**Comment by EmmadUsmani** [comment]
@alxnddr Thanks, yeah I can try to do that. Are there any specific cases you'd like to test regarding click behavior?

**Comment by alxnddr** [comment]
> Are there any specific cases you'd like to test regarding click behavior?
@EmmadUsmani it would be great to verify that clicks on circles on all levels pass correct data for example when interpolating URLs or setting dashboard filter values

**Pull request merged**